### PR TITLE
chore: add a config for running CI against the nightly driver MONGOSH-1473

### PR DESCRIPTION
**Pull request by lerouxb**
For use with https://spruce.mongodb.com/project/mongosh/settings/periodic-builds. Or any other way that you can specify an evergreen config file other than the default one.

### CODE REVIEW

**Comment by lerouxb** [comment]
Testing with `% ~/Downloads/evergreen create-version -m "nightly driver test (2)" --project=mongosh --path=.evergreen-nightly-driver.yml --active` which results in... it apparently not working https://parsley.mongodb.com/evergreen/mongosh_linux_unit_compile_ts_659bde7c9ccd4eddee20c2ed_24_01_08_11_37_32/0/task?bookmarks=0,9086

🤔

File: .evergreen.yml [hunk 1]
```diff
@@ -78,6 +78,7 @@ functions:
         env:
           NODE_JS_VERSION: ${node_js_version}
           DISTRO_ID: ${distro_id}
+          MONOGDB_VERSION_OVERRIDE: ${mongodb_version_override}
```
Comment (addaleax) [inline .evergreen.yml:81]: It's pretty unobvious that this refers to `mongodb` as in, the name of the literal `mongodb` driver npm package. Maybe `mongodb_driver_version_override` would be less ambiguous?

File: .evergreen/install-npm-deps.sh [hunk 2]
```diff
@@ -3,6 +3,14 @@ set -x
 
 npm ci --verbose
 
+echo "MONOGDB_DRIVER_VERSION_OVERRIDE:$MONOGDB_DRIVER_VERSION_OVERRIDE"
+
+# if MONOGDB_DRIVER_VERSION_OVERRIDE is set, then we want to replace the package version
+if [[ -n "$MONOGDB_DRIVER_VERSION_OVERRIDE" ]]; then
+  export REPLACE_PACKAGE="mongodb:$MONOGDB_DRIVER_VERSION_OVERRIDE"
+  npm run replace-package
+fi
```
Comment (addaleax) [inline .evergreen/install-npm-deps.sh:12]: From the evergreen log:

```
 [2024/01/08 13:03:58.391] Replacing mongodb in . with file:///data/mci/34be6bd4d487b3797e6cba6a385a9abf/src/nightly
```

That's not an actual file that exists, right? And `npm ls` still stays `mongodb@6.3.0`

Comment (lerouxb) [reply .evergreen/install-npm-deps.sh:12]:
```
~/mongo/mongosh % export REPLACE_PACKAGE=mongodb:nightly
~/mongo/mongosh % npm run replace-package

> replace-package
> node scripts/replace-package.js

Replacing mongodb in . with file:///Users/leroux.bodenstein/mongo/mongosh/nightly
Replacing mongodb in packages/arg-parser with file:///Users/leroux.bodenstein/mongo/mongosh/nightly
Replacing mongodb in packages/e2e-tests with file:///Users/leroux.bodenstein/mongo/mongosh/nightly
Replacing mongodb in packages/service-provider-core with file:///Users/leroux.bodenstein/mongo/mongosh/nightly
Replacing mongodb in packages/service-provider-server with file:///Users/leroux.bodenstein/mongo/mongosh/nightly
Replacing mongodb in packages/shell-api with file:///Users/leroux.bodenstein/mongo/mongosh/nightly
Replacing mongodb in packages/types with file:///Users/leroux.bodenstein/mongo/mongosh/nightly
~/mongo/mongosh % git status
On branch nightly-driver
Your branch is up to date with 'origin/nightly-driver'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   package-lock.json
	modified:   package.json
	modified:   packages/arg-parser/package.json
	modified:   packages/e2e-tests/package.json
	modified:   packages/service-provider-core/package.json
	modified:   packages/service-provider-server/package.json
	modified:   packages/shell-api/package.json
	modified:   packages/types/package.json

no changes added to commit (use "git add" and/or "git commit -a")
~/mongo/mongosh % git diff package.json
diff --git a/package.json b/package.json
index f9e95a34..e060057f 100644
--- a/package.json
+++ b/package.json
@@ -102,7 +102,7 @@
     "find-up": "^5.0.0",
     "husky": "^8.0.3",
     "mocha": "^10.2.0",
-    "mongodb": "^6.3.0",
+    "mongodb": "nightly",
     "mongodb-runner": "^5.4.6",
     "node-gyp": "^9.0.0",
     "nyc": "^15.1.0",
~/mongo/mongosh % npm i
npm WARN deprecated lodash.isarray@4.0.0: This package is deprecated. Use Array.isArray.
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated querystring@0.2.0: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated acorn-numeric-separator@0.3.6: acorn>=7.4 supports numeric separators
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142

> preinstall
> node scripts/sort-workspaces.js


> prepare
> husky install

husky - Git hooks installed

added 2637 packages, and audited 2666 packages in 48s

253 packages are looking for funding
  run `npm fund` for details

13 vulnerabilities (10 moderate, 3 high)

To address issues that do not require attention, run:
  npm audit fix

To address all issues possible (including breaking changes), run:
  npm audit fix --force

Some issues need review, and may require choosing
a different dependency.

Run `npm audit` for details.
~/mongo/mongosh % npm ls mongodb
mongosh@ /Users/leroux.bodenstein/mongo/mongosh
├─┬ @mongosh/arg-parser@0.0.0-dev.0 -> ./packages/arg-parser
│ ├─┬ @mongodb-js/devtools-connect@2.4.3
│ │ └── mongodb@6.3.0 deduped
│ └── mongodb@6.3.0 deduped
├─┬ @mongosh/e2e-tests@0.0.0-dev.0 -> ./packages/e2e-tests
│ └── mongodb@6.3.0 deduped
├─┬ @mongosh/service-provider-core@0.0.0-dev.0 -> ./packages/service-provider-core
│ └── mongodb@6.3.0 deduped
├─┬ @mongosh/service-provider-server@0.0.0-dev.0 -> ./packages/service-provider-server
│ └── mongodb@6.3.0 deduped
├─┬ @mongosh/shell-api@0.0.0-dev.0 -> ./packages/shell-api
│ └── mongodb@6.3.0 deduped
├─┬ @mongosh/types@0.0.0-dev.0 -> ./packages/types
│ └── mongodb@6.3.0 deduped
├─┬ mongodb-runner@5.4.6
│ └─┬ mongodb@5.8.1
│   └─┬ mongodb-client-encryption@2.9.0
│     └── mongodb@5.8.1 deduped
└── mongodb@6.3.0
```

yeah.

Comment (lerouxb) [reply .evergreen/install-npm-deps.sh:13]: Should be good now.

File: scripts/replace-package.js [hunk 3]
```diff
@@ -17,19 +16,28 @@ if (!parsed || !parsed.groups.from || !parsed.groups.to) {
...
+function resolveTag(from, to) {
+  return execSync(`npm dist-tag ls ${from}@${to} | awk -F ': ' '/^${to}/ {print \$2}'`).toString().trim();
+}
+
+const { from, to: _to } = parsed.groups;
+
+// npm install doesn't seem to do anything if you're updating a
+// package-lock.json file that already has the dep to a tag like nightly, but it
+// does do something if you change it to the exact version.
+const to = _to === 'nightly' ? resolveTag(from, _to) : _to;
```
Comment (addaleax) [inline scripts/replace-package.js:28]: At this point you might as well just make the script driver-specific? :)

File: scripts/replace-package.js [hunk 4]
```diff
@@ -17,19 +16,28 @@ if (!parsed || !parsed.groups.from || !parsed.groups.to) {
   throw new Error('Invalid format for REPLACE_PACKAGE');
 }
 
-const { from, to } = parsed.groups;
+function resolveTag(from, to) {
+  return execSync(`npm dist-tag ls ${from}@${to} | awk -F ': ' '/^${to}/ {print \$2}'`).toString().trim();
```
Comment (addaleax) [inline scripts/replace-package.js:20]:
```suggestion
  return execSync(`npm dist-tag ls '${from}@${to}' | awk -F ': ' '/^${to}/ {print \$2}'`).toString().trim();
```

**Review by addaleax — APPROVED** [review]

File: scripts/replace-package.js (hunk 3 above)
Comment (lerouxb) [reply scripts/replace-package.js:28]: The thought did cross my mind 😆

**Pull request merged**